### ISSUE TITLE
Running the subprocess as a different user

### ISSUE THREAD

**Issue by akoenig** [ahmetb/runsd#16]
@ahmetb As mentioned before, `runsd` is really nice and native support in Cloud Run would be a great addition, IMHO. We encountered one aspect though which we would like to challenge here.

In general, we're utilizing a non-privileged user within our container images. Under this user the actual service will be started. Due to the fact that `runsd` needs root permissions in order to change the `resolv.conf` we can't follow this approach when using `runsd`.

Do you think it would be "in scope" to pass the `uid` and `guid` to `runsd` so that it spawns the application by respecting the two values?

Curious to hear your thoughts on this and – as mentioned above – 🤞🏽 for having `runsd` natively in Cloud Run 🙂

**Comment by ahmetb** [comment]
I agree, we can do it.
Can you please fill out https://forms.gle/kCgEEiRqrmHhM65g6 if you are using it.
I couldn't see your name. :)

**Comment by akoenig** [comment]
I had already done it but now I filled it out again because something has changed in the project structure in the meantime and therefore the new information may also be relevant. 🙂

**Comment by erzz** [comment]
+1 - we really need to be running in a non-root context to be able to use this

**Comment by ahmetb** [comment]
@erzz 

- how would you all want to pass user id to this? number or name?
- do we need just uid, or both uid+gid?

**Comment by erzz** [comment]
@ahmetb 

I guess for flexibility (there are some very opinionated images out there!), all should be possible.

But if I take the majority of images they use user/group names rather than uid/gid so that would be a great start!

**Comment by ahmetb** [comment]
OK I think it's doable as well. I have an extra question (and these are because I actually never bothered to run as non-root before in a dockerfile):

Would you expect to use this as:

```
RUN useradd foo
USER foo
ENTRYPOINT ["/runsd", "--", "myapp"]
```

or:

```
RUN useradd foo
ENTRYPOINT ["/runsd", "-u=foo", "--", "myapp"]
```

I assume the latter but just wanna make sure!

Also, I might initially just add "user" support and leave "group" to a later version. Let me know if this is good to begin with @akoenig @erzz.

**Comment by ahmetb** [comment]
See #30 if you want to try this out by compiling from that branch.
I can merge soon and make a release if it's needed right away.

**Comment by erzz** [comment]
@ahmetb 

Hard to speak for anyone and everyone of course :) But I can describe at least our normal pattern and perhaps others can chime in if it is common practice for them too.

I'd say the majority of our images are multistage builds these days mostly for minimisation reasons.

Typically

```
FROM <node/java/whatever>:<whatever> AS builder
COPY /xyz /app
RUN yarn install or ./mvnw something yada yada

... other stuff ...

FROM <yarn/node/scratch/something>:<whatever>
RUN <somecreateusercommand>

ADD https://github.com/ahmetb/runsd/releases/download/v0.0.0-rc.12/runsd /bin/runsd
RUN chmod +x /bin/runsd

COPY --chown=user:group --from=builder /thing/built/in/stage1 .
ENV HOST 0.0.0.0

USER user

CMD ["runsd", "--", "yarn", "start"]
```

In almost every case of "ours" if a group is created its almost cosmetic :) The user is the thing we run as - so I think that would be a case that would satisfy 99.8% of our cases :)

For "us" it would be much nicer to use 

```
RUN useradd foo
USER foo
ENTRYPOINT ["/runsd", "--", "myapp"]
```

as anything that doesn't add complexity to or juggling of the entrypoint command will be nicer and more portable between images built in different ways

**Comment by ahmetb** [comment]
> For "us" it would be much nicer to use
> 
> 
> ```
> RUN useradd foo
> USER foo
> ENTRYPOINT ["/runsd", "--", "myapp"]
> ```

the problem is I am not sure if we can re-escalate to `root` (runsd needs it) after we adjust the user as `foo` from that step on.

**Comment by erzz** [comment]
No probs - a "nice-to-have" only.

#30 solution will be good enough for sure

**Comment by akoenig** [comment]
From our side: Passing the user via argument is perfectly fine. Thanks a lot, @ahmetb

**Comment by ahmetb** [comment]
Released v0.0.0-rc.13 please try it out.
